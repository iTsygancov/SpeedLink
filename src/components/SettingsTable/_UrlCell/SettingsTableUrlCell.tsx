import { Input } from "@/components/ui/input";
import { TableCell } from "@/components/ui/table";
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger
} from "@/components/ui/tooltip";
import { useClipboard } from "@/lib/hooks/useClipboard";
import { Command } from "@/types";
import { Check, Clipboard } from "lucide-react";

type SettingsTableUrlCellProps = {
  handleChangeShortcut: (
    event: React.ChangeEvent<HTMLInputElement>,
    itemIndex: number
  ) => void;
  item: Command;
  itemIndex: number;
};

const SettingsTableUrlCell = ({
  handleChangeShortcut,
  item,
  itemIndex
}: SettingsTableUrlCellProps) => {
  const { isCopied, copy } = useClipboard();

  return (
    <TableCell className='break-all p-0 px-4 py-2'>
      {item.canEdit ? (
        <Input
          value={item.url}
          name='url'
          onChange={(event) => handleChangeShortcut(event, itemIndex)}
        />
      ) : (
        <div className='flex items-center justify-between gap-8'>
          <a
            className='hover:underline'
            href={item.url}
            target='_blank'
            rel='noreferrer noopener'
          >
            {item.url}
          </a>
          <TooltipProvider>
            <Tooltip open={isCopied}>
              <TooltipTrigger className='p-2' onClick={() => copy(item.url)}>
                {isCopied ? <Check size={16} /> : <Clipboard size={16} />}
              </TooltipTrigger>
              <TooltipContent>
                <p>Copied to clipboard</p>
              </TooltipContent>
            </Tooltip>
          </TooltipProvider>
        </div>
      )}
    </TableCell>
  );
};

export default SettingsTableUrlCell;
