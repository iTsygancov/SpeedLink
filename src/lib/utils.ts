import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function generateAlphanumericKeysArray() {
  const keysArray = [];

  // Generate lowercase alphabetic keys
  for (let i = 65; i <= 90; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  // Generate numeric keys
  for (let i = 48; i <= 57; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  return keysArray;
}
