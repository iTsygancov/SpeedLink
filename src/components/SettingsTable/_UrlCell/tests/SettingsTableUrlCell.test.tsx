import SettingsTableUrlCell, {
  SettingsTableUrlCellProps
} from "../SettingsTableUrlCell";
import { initialProps } from "./SettingsTableUrlCell.data";
import { expectToMatchSnapshot, renderTree } from "@/test/utils";
import { describe, it } from "vitest";

const testComponentWithProps = (props: SettingsTableUrlCellProps) => {
  expectToMatchSnapshot(renderTree(<SettingsTableUrlCell {...props} />));
};

describe("SettingsTableUrlCell", () => {
  it("should render with empty url", () => {
    testComponentWithProps(initialProps);
  });

  it("should render with filled url and copy button", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        url: "https://example.com"
      }
    });
  });

  it("should render with empty input field", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        canEdit: true
      }
    });
  });

  it("should render with filled input field", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        canEdit: true,
        url: "https://example.com"
      }
    });
  });
});
