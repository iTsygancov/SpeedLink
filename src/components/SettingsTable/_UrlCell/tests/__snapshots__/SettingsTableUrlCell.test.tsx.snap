// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsTableUrlCell > should render with empty input field 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 break-all p-0 px-4 py-2"
>
  <input
    className="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
    name="url"
    onChange={[Function]}
    value=""
  />
</td>
`;

exports[`SettingsTableUrlCell > should render with empty url 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 break-all p-0 px-4 py-2"
>
  <div
    className="flex items-center justify-between gap-8"
  >
    -
  </div>
</td>
`;

exports[`SettingsTableUrlCell > should render with filled input field 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 break-all p-0 px-4 py-2"
>
  <input
    className="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
    name="url"
    onChange={[Function]}
    value="https://example.com"
  />
</td>
`;

exports[`SettingsTableUrlCell > should render with filled url and copy button 1`] = `
<td
  className="align-middle [&:has([role=checkbox])]:pr-0 break-all p-0 px-4 py-2"
>
  <div
    className="flex items-center justify-between gap-8"
  >
    <a
      className="hover:underline"
      href="https://example.com"
      rel="noreferrer noopener"
      target="_blank"
    >
      https://example.com
    </a>
    <button
      className="p-2"
      data-state="closed"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onPointerDown={[Function]}
      onPointerLeave={[Function]}
      onPointerMove={[Function]}
    >
      <svg
        className="lucide lucide-clipboard"
        fill="none"
        height={16}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
        viewBox="0 0 24 24"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="4"
          rx="1"
          ry="1"
          width="8"
          x="8"
          y="2"
        />
        <path
          d="M16 4h2a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2"
        />
      </svg>
    </button>
  </div>
</td>
`;
